cycles-ledger/src/storage.rs

```diff
@@ -766,6 +766,19 @@ pub fn transfer(
 ) -> Result<Nat, TransferFromError> {
     use TransferFromError::*;
 
+    let transaction = Transaction {
```

Comment:
What's the motivation for performing this check first?

Reply:
When you submit the transaction first time around, it updates your balance and then makes an async call. If at this state you resubmit transaction you might not have enough balance to pass the balance check (or expected_allowance might not match in case of approve). In such case you will get an InsufficientFunds error instead of DuplicateTransaction.
On the other hand you need to update the balance before that async call, otherwise you enter all sorts of double spending scenarios. (sorry, I should have put this explanation in the description, updating now).

Reply:
But yes, ultimately the transaction wouldn't go through anyway. It is just about user experience/returning correct error.